## Customer Churn Prediction using Logistic Regression

### Notebook 02 – checking the values by analysing the missing values and duplicated and shape 

### Introduction

In the previous notebook, we explored the dataset and understood its structure, available features, and target variable.

In this notebook, we will clean the dataset by identifying duplicate records, handling missing values, removing unnecessary features, correcting data types, and preparing the data for Exploratory Data Analysis (EDA) and model training.

### Learning Objectives

By the end of this notebook, you will be able to:

- Import the required libraries
- Load the dataset
- Create a working copy
- Check duplicate records
- Identify missing values
- Remove unnecessary columns
- Prepare a clean dataset for the next notebook

### Step 1 – Import Required Libraries

The first step in every notebook is to import the required Python libraries.

Keeping each notebook self-contained ensures it can be executed independently without relying on previous notebooks. This makes the project easier to understand, maintain, and share.

In [2]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np

### Step 2 – Load the Dataset

After importing the required libraries, we load the customer churn dataset into a Pandas DataFrame.

This dataset will be used throughout this notebook for cleaning and preprocessing tasks.

In [3]:
df = pd.read_excel(r"D:\machine learning\machine_learning\logistic_regression_project\churn_prediction\data\raw _data\Telco_customer_churn.xlsx")
df

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,2569-WGERO,1,United States,California,Landers,92285,"34.341737, -116.539416",34.341737,-116.539416,Female,...,Two year,Yes,Bank transfer (automatic),21.15,1419.4,No,0,45,5306,NaN
7039,6840-RESVB,1,United States,California,Adelanto,92301,"34.667815, -117.536183",34.667815,-117.536183,Male,...,One year,Yes,Mailed check,84.80,1990.5,No,0,59,2140,NaN
7040,2234-XADUH,1,United States,California,Amboy,92304,"34.559882, -115.637164",34.559882,-115.637164,Female,...,One year,Yes,Credit card (automatic),103.20,7362.9,No,0,71,5560,NaN
7041,4801-JZAZL,1,United States,California,Angelus Oaks,92305,"34.1678, -116.86433",34.167800,-116.864330,Female,...,Month-to-month,Yes,Electronic check,29.60,346.45,No,0,59,2793,NaN


### Step 3 – Create a Working Copy of the Dataset

Before making any modifications, we create a copy of the original dataset.

Working on a copy protects the original data and allows us to restart the preprocessing process whenever necessary.

In [4]:
df_clean = df.copy()

### Step 4 — Check for Duplicate Records

Duplicate records can negatively impact the performance of a machine learning model because the same information is counted multiple times during training.

Before removing any duplicates, it is important to determine whether duplicate records actually exist in the dataset. If duplicates are found, we will remove them to ensure that each customer is represented only once.

In [5]:
df_clean.duplicated().sum()

np.int64(0)

### Observation

The dataset does not contain any duplicate records.

Since no duplicate entries were found, there is no need to perform duplicate removal. We can proceed to the next data cleaning step.

### Step 5 — Check for Missing Values

Missing values are one of the most common problems encountered in real-world datasets.

Before building a machine learning model, it is essential to identify columns containing missing values, as they can affect data quality and model performance. Once identified, we can decide whether to remove or impute these missing values based on their significance.

In [6]:
df_clean.isnull().sum()

CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges           0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

### Observation

the dataset looks like does not have missing value but in the churn reason there was 5174 missing value 

In the next step, we will calculate the percentage of missing values to better understand their impact and determine the most appropriate handling strategy.

### Step 6 — Calculate the Percentage of Missing Values

While the total number of missing values is useful, calculating their percentage provides better insight into the severity of missing data.

Columns with a very high percentage of missing values may need to be removed, whereas columns with only a few missing values can often be treated using suitable imputation techniques.

In [7]:
missing_values = pd.DataFrame({
    "Missing Values": df_clean.isnull().sum(),
    "Percentage (%)": (df_clean.isnull().sum() / len(df_clean)) * 100
})

missing_values[missing_values["Missing Values"] > 0].sort_values(
    by="Percentage (%)",
    ascending=False
)

,Missing Values,Percentage (%)
Churn Reason,5174,73.463013


### Observation

The missing value report helps identify which features require preprocessing.

Based on the percentage of missing values, we can decide whether to remove a feature or apply an appropriate imputation technique during data cleaning.

### Step 7 — Inspect Feature Data Types

Understanding the data type of each feature is an essential step before preprocessing.

Machine learning algorithms require numerical input; therefore, identifying numerical and categorical features allows us to apply the correct preprocessing techniques, such as encoding categorical variables or converting data types where necessary.

In [8]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   object 
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   object 
 3   State              7043 non-null   object 
 4   City               7043 non-null   object 
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   object 
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   object 
 10  Senior Citizen     7043 non-null   object 
 11  Partner            7043 non-null   object 
 12  Dependents         7043 non-null   object 
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   object 
 15  Multiple Lines     7043 non-null   object 
 16  Internet Service   7043 